planner/core/memtable_predicate_extractor.go

```diff
@@ -466,3 +476,95 @@ func (e *ClusterLogTableExtractor) Extract(
 	e.Patterns = patterns
 	return remained
 }
+
+// MetricTableExtractor is used to extract some predicates of metric_schema tables.
+type MetricTableExtractor struct {
+	extractHelper
+	// SkipRequest means the where clause always false, we don't need to request any component
+	SkipRequest bool
+	// StartTime represents the beginning time of metric data.
+	StartTime int64
+	// EndTime represents the ending time of metric data.
+	EndTime int64
+	// LabelConditions represents the label conditions of metric data.
+	LabelConditions map[string]set.StringSet
+	quantile        set.StringSet
+}
+
+// Extract implements the MemTablePredicateExtractor Extract interface
+func (e *MetricTableExtractor) Extract(
+	ctx sessionctx.Context,
+	schema *expression.Schema,
+	names []*types.FieldName,
+	predicates []expression.Expression,
+) []expression.Expression {
+	// Extract the `quantile` columns
+	remained, skipRequest, quantile := e.extractCol(schema, names, predicates, "quantile", true)
+	e.SkipRequest = skipRequest
+	if e.SkipRequest {
+		return nil
+	}
+	e.quantile = quantile
+
+	// Extract the `time` columns
+	remained, startTime, endTime := e.extractTimeRange(ctx, schema, names, remained, "time")
+	if endTime == 0 {
+		endTime = math.MaxInt64
+	}
+	e.StartTime = startTime
+	e.EndTime = endTime
+	e.SkipRequest = startTime > endTime
+
+	if e.SkipRequest {
+		return nil
+	}
+
+	// Extract the label columns.
+	for _, name := range names {
+		switch name.ColName.L {
+		case "quantile", "time", "value":
+			continue
+		}
+		var values set.StringSet
+		remained, skipRequest, values = e.extractCol(schema, names, remained, name.ColName.L, false)
+		if skipRequest {
+			e.SkipRequest = skipRequest
+			return nil
+		}
+		if len(values) == 0 {
+			continue
+		}
+		if e.LabelConditions == nil {
+			e.LabelConditions = make(map[string]set.StringSet)
+		}
+		e.LabelConditions[name.ColName.L] = values
+	}
+	return remained
+}
+
+// GetQuantiles gets the quantiles of metric query.
+func (e *MetricTableExtractor) GetQuantiles() ([]float64, error) {
+	if len(e.quantile) == 0 {
+		return []float64{0}, nil
+	}
+	quantiles := make([]float64, 0, len(e.quantile))
+	for k := range e.quantile {
+		quantile, err := strconv.ParseFloat(k, 64)
+		if err != nil {
+			return nil, err
+		}
+		quantiles = append(quantiles, quantile)
+	}
+	sort.Float64s(quantiles)
+	return quantiles, nil
+}
+
+// GetQueryRangeTime gets the metric query time range.
+func (e *MetricTableExtractor) GetQueryRangeTime(sctx sessionctx.Context) (start, end time.Time, step time.Duration) {
+	start = e.convertToTime(e.StartTime)
+	end = e.convertToTime(e.EndTime)
+	if start.After(end) {
+		start = end
+	}
+	return start, end, time.Second * time.Duration(sctx.GetSessionVars().MetricSchemaStep)
+}
```

Comment:
Prefer to move those two functions to `executor` package.

Reply:
Explain will use this function, Move to executor will have circle import problem.